src/Avalonia.Styling/Styling/Selector.cs

```diff
@@ -66,16 +69,39 @@ public SelectorMatch Match(IStyleable control, bool subscribe = true)
                 }
                 else if (match.Result == SelectorMatchResult.Sometimes)
                 {
-                    inputs.Add(match.Activator);
+                    Debug.Assert(match.Activator != null);
+
+                    if (inputs != null)
```

Comment:
It might be worth to have a struct wrapping a single value and a list with a list-like API. Like [StringValues](https://github.com/dotnet/extensions/blob/master/src/Primitives/src/StringValues.cs), but mutable. That way this code will be moved inside of the struct and API consumer would still just call Add and let the struct to create the list automatically when needed.

Reply:
I was thinking about having such thing as well, will investigate. Bad part is that it is annoying to provide more than one element of storage and I will need to figure how to deal with `StyleActivator.And` which still needs `IList` to pass to Rx. Although in such hot code path we might create our own observers for `And` and `Or` , this should also reduce footprint a bit.

Reply:
Well, it could expose `Single` and `List` properties, so [this line](https://github.com/AvaloniaUI/Avalonia/pull/3474/files#diff-a7d2f60c59bdc29bde56fece096275b5R97) would just check for `IsSingle` instead and then get the single element or the list if needed.

The struct could even be refonly to clearly state that it's intended to be a temporary wrapper.